kivy/loader.py

```diff
@@ -122,6 +124,10 @@ def __init__(self):
         self._start_wanted = False
         self._trigger_update = Clock.create_trigger(self._update)
 
+        if platform == 'android':
+            import certifi
+            environ['SSL_CERT_FILE'] = certifi.where()
```

Comment:
I didn't follow the conversation, but can you summarize why this variable is set?

Reply:
Ah, I see https://github.com/kivy/python-for-android/issues/1827#issuecomment-500028459.

I'll trust @tshirtman judgement here, but I am a little weary of overriding this variable from within kivy. I.e. how do we know someone may not have set it to some specific value? Is this a kivy specific issue that we need to set it like this only from kivy and not elsewhere? What about non-kivy apps on android? Maybe py4a should set this?

And if this is a global issue, maybe this needs to be set in kivy's `__init__`?

Reply:
Without setting this variable: a kivy application works out of the box on Linux. However, on physical deployment a black screen is rendered, and no error messages are reported in buildozer nor in logcat. Leading to an extremely difficult problem to debug for kivy users (literally the hardest problem I have ever seen; harder than any problem I have seen even in mathematics.)

This setting this variable causes consistent kivy behavior across Linux, and Android. Also it maintains the original intent of the kivy API on these access points.

Reply:
> I didn't follow the conversation, but can you summarize why this variable is set?

Sorry, I misspoke. The AsyncImage issue only caused AsyncImages to fail to load, but did not prevent the application from rendering. UrlRequest caused the black screen.

Reply:
> I didn't follow the conversation, but can you summarize why this variable is set?

Addressing this issue: https://github.com/kivy/python-for-android/issues/1827

Reply:
Hm, if it's already set, should we override it? That's a fair point, i don't think it's likely to be an issue in practice, but we could check and only define it if it doesn't exist already, os.environ acting like a dict certainly has a setdefault method that would make this convenient.

Reply:
Yeah, and in the long-term, @andnovar [seems to think](https://github.com/kivy/python-for-android/issues/1827#issuecomment-500078114) he'd do it in py4a. But until then it should be ok to use `setdefault`.

Reply:
> Ah, I see [kivy/python-for-android#1827 (comment)](https://github.com/kivy/python-for-android/issues/1827#issuecomment-500028459).
> 
> I'll trust @tshirtman judgement here, but I am a little weary of overriding this variable from within kivy. I.e. how do we know someone may not have set it to some specific value? Is this a kivy specific issue that we need to set it like this only from kivy and not elsewhere? What about non-kivy apps on android? Maybe py4a should set this?
> 
> And if this is a global issue, maybe this needs to be set in kivy's `__init__`?

On my Android

~$ echo SSL_CERT_FILE

Is empty.

Environment variables set in Python are not persistent, and are isolated to their process. A kivy application could set this value, thus overwriting the solution here, but doing so would break the contract given by this fix. So existing applications that have this statement in their main.py are unaffected by the fix. And new applications do not need to include the statement, although if they do, it won't cause any issues.

Specifically the issue solved occurs at urllib.request.Request in the kivy.loader.LoaderBase class, within the kivy framework.

I didn't look at p4a. But p4a should set this if it uses the urllib.request.Request to make https requests and fails. I believe that the causes of the issue occurs around the transition from Python 2, to Python 3, with porting urllib to Python 3, along with (at around the same time) the web adopting the HTTPS standard over HTTP.

I don't think this issue is a global issue, because it depends on how the request is implemented. The statement in __init__ does fix UrlRequest, for example. Or UrlRequest can just directly ask for the cacert.pem location since it uses the urllib.request.urlrequest interface. AsyncImage cannot directly ask for cacert.pem, and must use the environment variable because it uses the urllib.request.Request interface. Changing this, by refactoring the request code in LoaderBase, could solve the issue without having to set the environment variable, although I am leery of refactoring code that otherwise totally works in every other respect.

Putting it in __init__ does work, and was the originally submitted solution; because I couldn't get the build to include certifi in the APK (which was what I was actually trying to do at first.) Placing it in __init__ is better for kivy maintainers because it fully abstracts the problem into a single statement, but is worst for the API because a behavioral inconsistency still exists in that case: Linux applications would not require import kivy to make requests successfully, but Android applications would require import kivy. If the user doesn't know this they will still run into a similar problem that I faced, although that problem might be easier to deal with for new users with the kivy/__init__.py solution since a lot of existing tutorials on kivy have the user import kivy anyways. However, the current fix implements a consistent application behavior across all platforms, without having to write platform specific code, and I feel that this fully solves the problems that I, and many others have faced. What I mean is that, in all cases, on Linux you would not need to import kivy, but in the case of placing the fix in kivy/__init__.py Android HTTPS applications would have to have the extra import kivy statement; thus, platform specific code for kivy users.

TL;DR: this version completely solves the problem of needing platform specific code for kivy users in making https requests; the __init__ version attempts to lessen the severity of the problem, but doesn't really fully solve the problem, and still results in platform specific code. It was my original submission, because I thought it was the best we could do. But with tshirtmans help and knowledge, we were able to arrive at maintaining a 100% platform independent behavioral consistency. So I am super biased in favor of this version.

Reply:
> Ah, I see [kivy/python-for-android#1827 (comment)](https://github.com/kivy/python-for-android/issues/1827#issuecomment-500028459).
> 
> I'll trust @tshirtman judgement here, but I am a little weary of overriding this variable from within kivy. I.e. how do we know someone may not have set it to some specific value? Is this a kivy specific issue that we need to set it like this only from kivy and not elsewhere? What about non-kivy apps on android? Maybe py4a should set this?
> 
> And if this is a global issue, maybe this needs to be set in kivy's `__init__`?

Sorry the comments are out of order. The TL;DR pertains to your most recent comment.

The comments above pertain to your second most recent comment. For some reason I could not see your most recent comment until, basically, just now.

Reply:
> Ah, I see [kivy/python-for-android#1827 (comment)](https://github.com/kivy/python-for-android/issues/1827#issuecomment-500028459).
> 
> I'll trust @tshirtman judgement here, but I am a little weary of overriding this variable from within kivy. I.e. how do we know someone may not have set it to some specific value? Is this a kivy specific issue that we need to set it like this only from kivy and not elsewhere? What about non-kivy apps on android? Maybe py4a should set this?
> 
> And if this is a global issue, maybe this needs to be set in kivy's `__init__`?

We could check whether SSL_CERT_FILE is not empty before setting, as suggested by AndreMiras in kivy/python-for-android#1827. But it is empty by default according to Sirfanas.

Writing without checking shouldn't break anything, unless some other part of the framework is a determinate and setting it to a different value.

Reply:
The only way i would see it not empty, would be that the user somehow manipulated the env themselves before importing kivy, and in such case we would be overriding their choice.

using setdefault instead of `=` avoids this corner case.

```
>>> import os
>>> os.environ.get('TEST')
>>> os.environ.setdefault('TEST', 'test')
'test'
>>> os.environ.get('TEST')
'test'
>>> os.environ.setdefault('TEST', 'test1')
'test'
>>> os.environ.get('TEST')
'test'
```

Reply:
Hey guys. I've pushed the code using setdefault in place of the assignment statement, in the loader module.